packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -63,31 +70,121 @@ const isRwLoggingError = (x: any): x is RwLoggingError =>
     typeof x?.addTelemetryProperties === "function" && isILoggingError(x);
 
 /**
- * Annotate the given error object with the given logging props
- * @returns The same error object passed in if possible, with telemetry props functionality mixed in
+ * Mixes in the given properties, accessible henceforth via ILoggingError.getTelemetryProperties
+ * @returns the same object that was passed in, now also implementing ILoggingError.
  */
-export function annotateError(
-    error: unknown,
+export function mixinTelemetryProps<T extends Record<string, unknown>>(
+    error: T,
     props: ITelemetryProperties,
-): ILoggingError {
+): T & ILoggingError {
+    assert(isRegularObject(error) && !Object.isFrozen(error), "Cannot mixin Telemetry Props");
+
     if (isRwLoggingError(error)) {
         error.addTelemetryProperties(props);
         return error;
     }
 
-    if (isRegularObject(error)) {
-        // Even though it's not exposed, fully implement IRwLoggingError for subsequent calls to annotateError
-        const loggingError = error as RwLoggingError;
-
-        const propsForError = {...props};
-        loggingError.getTelemetryProperties = () => propsForError;
-        loggingError.addTelemetryProperties =
-            (newProps: ITelemetryProperties) => { copyProps(propsForError, newProps); };
-        return loggingError;
+    // Even though it's not exposed, fully implement RwLoggingError for subsequent calls to mixingTelemetryProps
+    const loggingError = error as T & RwLoggingError;
+
+    const propsForError = {...props};
+    loggingError.getTelemetryProperties = () => propsForError;
+    loggingError.addTelemetryProperties =
+        (newProps: ITelemetryProperties) => { copyProps(propsForError, newProps); };
+
+    return loggingError;
+}
+
+/**
+ * Take an unknown error object and extract certain known properties to be included in a new error object.
+ * The stack is preserved, along with any safe-to-log telemetry props.
+ * @param error - An error that was presumably caught, thrown from unknown origins
+ * @param newErrorFn - callback that will create a new error given the original error's message
+ * @returns A new error object "wrapping" the given error
+ */
+ export function wrapError<T>(
+    error: any,
+    newErrorFn: (m: string) => T,
+): T {
+    const {
+        message,
+        stack,
+    } = extractLogSafeErrorProperties(error);
+    const props = isILoggingError(error) ? error.getTelemetryProperties() : {};
+
+    const newError = newErrorFn(message);
+    mixinTelemetryProps(newError as ExtensibleObject<T>, props);
```

Comment:
you do a lot of casting for this function, which breaks all type safety, why bother, and not just have mixinTelemetryProps take T, as you already do the runtime check in the funciton itself

Reply:
Yeah great point, thanks.